alf/environments/suite_carla.py

```diff
@@ -689,8 +731,80 @@ def transform_to_geolocation(self, location: carla.Location):
         return np.array([loc.latitude, loc.longitude, loc.altitude])
 
 
+class NavigationSensor(SensorBase):
+    """Generating future waypoints on the route.
+
+    Note that the route is fixed (not change based on current vehicle location)
+    """
+
+    WINDOW = 5
+
+    def __init__(self, parent_actor, alf_world: World):
+        """
+        Args:
+            parent_actor (carla.Actor): the parent actor of this sensor
+            alf_world (World):
+        """
+        super().__init__(parent_actor)
+        self._alf_world = alf_world
+        self._future_indices = np.array([1, 3, 10, 30, 100, 300, 1000, 3000])
+
+    def set_destination(self, destination):
+        """Set the navigation destination.
+
+        Args:
+            destination (carla.Location):
+        """
+        start = self._alf_world.get_vehicle_location(self._parent.id)
+        self._route = self._alf_world.trace_route(start, destination)
+        self._waypoints = np.array([[
+            wp.transform.location.x, wp.transform.location.y,
+            wp.transform.location.z
+        ] for wp, _ in self._route])
+        self._road_option = np.array(
+            [road_option for _, road_option in self._route])
+        self._nearest_index = 0
+
+    def observation_spec(self):
+        return alf.TensorSpec([len(self._future_indices), 3])
+
+    def observation_desc(self):
+        return ("Positions of the %s future locations in the route." % len(
+            self._future_indices))
+
+    def get_current_observation(self, current_frame):
+        """Get the current observation.
```

Comment:
More descriptive? Might copy `observation_desc` here. Also explain the observation spec.